### PR TITLE
fixes #2434 - foreman-debug script

### PR DESCRIPTION
**Pull request by lzap**
foreman-debug -h

```
Collects configuration and log data for Foreman, Smart Proxies, backend
services and system information while removing security information like
passwords, tokens and keys.

This program can be used on Foreman instance, Smart Proxy instances or
backend services separatly.

OPTIONS:
  -d DIR  Directory to place the tarball in (default /tmp/foreman-XYZ)
  -g      Skip generic info (CPU, memory, firewall etc.)
  -a      Do not generate a tarball from the resulting directory
  -m      Maximum lines to keep for each file (default 5000)
  -j PRG  Filter with provided program when creating a tarball
  -p      Additionally print all passwords which are being filtered
  -q      Quiet mode
  -v      Verbose mode
  -h      Shows this message
```

### CODE REVIEW

File: script/foreman-debug [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+[ $DEBUG -eq 0 -a $EUID -ne 0 ] && error "This script must be run as root" && exit 1
+
+# determine distribution family
+if [ -f /etc/debian_version ]; then
+    OS=debian
+    OS_RELEASE=$(lsb_release -sr)
+elif [ -f /etc/redhat-release ]; then
+    OS=redhat
+    OS_RELEASE=$(lsb_release -sr)
```
Comment (domcleal) [inline script/foreman-debug:157]: We might need to fallback on `rpm -q --whatprovides redhat-release` here, as it's common for LSB utilities to not be installed on RH derivatives.

File: script/foreman-debug [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+# GENERIC ARTIFACTS
+
+if [ $NOGENERIC -eq 0 ]; then
+  printv "Collecting generic system information"
+  add_cmd "uname -a" "uname"
+  add_cmd "cat /proc/cpuinfo" "cpuinfo"
+  add_cmd "cat /proc/meminfo" "meminfo"
+  add_cmd "ulimit -a" "ulimit"
+  add_cmd "lsmod" "lsmod"
+  add_cmd "iptables -L" "iptables"
```
Comment (domcleal) [inline script/foreman-debug]: Add -v here too and perhaps -n, it gives all columns and details of rules.

File: script/foreman-debug [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+  add_cmd "ifconfig -a" "ifconfig"
+  add_cmd "route -n" "route"
+  add_cmd "netstat -tulpn" "netstat"
```
Comment (domcleal) [inline script/foreman-debug]: These three have gone away in Fedora 18 by default, so `ip a`, `ip r` and `ss -tulpn` might be more portable.

File: script/foreman-debug [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+printv "Collecting Foreman-related information"
+add_cmd "gem list" "gem_list"
+add_cmd "bundle --local --gemfile=/usr/share/foreman/Gemfile" "bundle_list"
+add_cmd "facter" "facts"
+add_files /etc/foreman{,-proxy}/*.yaml /var/log/foreman{,-proxy}/*.log
+add_files /usr/share/foreman/Gemfile*
+add_files /etc/dhcp/*.conf /var/lib/dhcp/*.leases
+add_files /etc/xinetd.d/tftp
+add_cmd "find /var/lib/tftpboot" "tftpboot_tree"
```
Comment (domcleal) [inline script/foreman-debug]: Add /tftpboot for RHEL 5 and /srv/tftp for Debian.

Comment (domcleal) [inline script/foreman-debug]: Perhaps add `-exec ls -ld {} +` too, as a common problem is zero length files in boot/.

File: script/foreman-debug [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+add_files /etc/named.conf /var/log/named.log 
+add_files /var/named/* /var/named/{data,dynamic,slaves}/*
+add_cmd "virsh list" "virsh_list"
+add_files /etc/libvirt/* /etc/libvirt/storage/* /etc/libvirt/qemu/* /etc/libvirt/qemu/networks
+add_files /var/lib/pgsql/data/*.conf
+add_files /etc/puppet/*.conf /etc/puppet/rack/* /etc/puppet/manifests/* /var/log/puppet/*.log
+add_files /var/lib/puppet/ssl/certs/$(hostname -f).pem /var/lib/puppet/ssl/certs/ca.pem
+add_files /etc/puppet/ssl/ca/inventory.txt
```
Comment (domcleal) [inline script/foreman-debug]: This should be `/var/lib/puppet/ssl/ca/inventory.txt` on a default packaged install.  I know yours is a bit funky :)

File: script/foreman-debug [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+add_cmd "find /etc/puppet/modules" "puppet_manifests_tree"
+add_files /etc/{httpd,apache2}/conf/*
+add_files /etc/{httpd,apache2}/conf.d/*
+add_cmd "echo 'select id,name,value from settings' | su postgres -c 'psql foreman'" "foreman_settings_table"
+
+add_files /etc/{sysconfig,default}/foreman{,-proxy}
+add_files /etc/{sysconfig,default}/dhcp*
+add_files /etc/{sysconfig,default}/named
```
Comment (domcleal) [inline script/foreman-debug]: This is `/etc/default/bind*` on Debian.

File: script/foreman-debug [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+add_files /etc/{sysconfig,default}/libvirtd
```
Comment (domcleal) [inline script/foreman-debug]: `libvirt*` would catch the `libvirt-bin` (I think) file on Debian too.

File: script/foreman-debug [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+add_files /etc/{sysconfig,default}/pgsql
```
Comment (domcleal) [inline script/foreman-debug]: This is `/etc/default/postgresql` on Debian I think.

File: script/foreman-debug [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+
+qprintf "\n\n"
+qprintf "%10s %s\n" "HOSTNAME:" "$(hostname -f 2>/dev/null)"
```
Comment (domcleal) [inline script/foreman-debug]: Do we capture these lines in the tarball too?

**Comment by domcleal** [comment]
That looks good @lzap!  I've added some comments about specific files/commands, but I think we can merge once they're fixed.

Do you think should package this at /usr/sbin/foreman-debug or similar?

File: script/foreman-debug (hunk 7 above)
Comment (lzap) [inline script/foreman-debug]: Not following this one

File: script/foreman-debug (hunk 8 above)
Comment (lzap) [inline script/foreman-debug]: Oh right there is no default file for postgres.

File: script/foreman-debug (hunk 9 above)
Comment (lzap) [inline script/foreman-debug]: We were not, few missed.

Ok I fixed all your comments, many thanks. Pushing update.

Comment (lzap) [inline script/foreman-debug]: We were not, few missed.

Ok I fixed all your comments, many thanks. Pushing update.

File: script/foreman-debug (hunk 7 above)
Comment (lzap) [inline script/foreman-debug]: Got it! All fixed now.

**Comment by domcleal** [comment]
Thanks @lzap, merged as bd79f51.

**Pull request closed without merging**